### PR TITLE
universal_addresses: configure sizes according to the used modules

### PR DESCRIPTION
**Pull request by OlegHahm**
The maximum size of a universal address should be dependent on the used modules. The size of the universal address table is dependent on the size of the FIB (if used).

~~Depends on #3655.~~

### CODE REVIEW

**Comment by OlegHahm** [comment]
rebased to current master

**Comment by BytesGalore** [comment]
I would say please squash.

File: sys/universal_address/universal_address.c [hunk 1]
```diff
@@ -28,7 +31,11 @@
 /**
  * @brief Maximum number of entries handled
  */
+#ifdef MODULE_FIB
+#define UNIVERSAL_ADDRESS_MAX_ENTRIES (2 * FIB_MAX_FIB_TABLE_ENTRIES)
```
Comment (BytesGalore) [inline sys/universal_address/universal_address.c]: just a comment (no need to change something)
Twice the `FIB_MAX_FIB_TABLE_ENTRIES` is safe but only necessary if all addresses (destination and next-hop) used in for the FIB entries are unique.
Multiple equal addresses only occupy a single memory for the address.

**Comment by OlegHahm** [comment]
fixed an issue with doxygen

**Comment by BytesGalore** [comment]
then please squash and then lets trigger travis

**Comment by OlegHahm** [comment]
squashed and asked Mr. T

File: sys/include/universal_address.h [hunk 2]
```diff
@@ -28,7 +28,18 @@ extern "C" {
 #include <stdint.h>
 #include <stdlib.h>
 
-#define UNIVERSAL_ADDRESS_SIZE (16)         /**< size of the used addresses in bytes */
+/** @brief  size of the used addresses in bytes */
+/* determine the widest possible address type */
+#ifndef UNIVERSAL_ADDRESS_SIZE
+#define UNIVERSAL_ADDRESS_SIZE (0)       /* rather senseless default, should
```
Comment (BytesGalore) [inline sys/include/universal_address.h]: Mr. GCC for x86 doesn't like to use `0` size for the 
``` C
uint8_t address[UNIVERSAL_ADDRESS_SIZE]; /**< The generic address data */
``` 
in line 50

Comment (BytesGalore) [inline sys/include/universal_address.h]: maybe we could use a default size of 1 and add:
``` C
#warning "The UNIVERSAL_ADDRESS_SIZE has not been set yet. Using default of 1."
#define UNIVERSAL_ADDRESS_SIZE (1)       /* rather senseless default, should
```
or similar

Comment (OlegHahm) [inline sys/include/universal_address.h]: Well, the `#warning` is gcc specific and would trigger an error. Actually, I'm not sure, but I think I thought about this problem while creating the PR and the conclusion might have been that actually rising an error if this module is included without setting the value for its size **is** an error.

Comment (OlegHahm) [inline sys/include/universal_address.h]: Hence, I think the solution is to let the unittest set this variable. Do you agree?

**Comment by BytesGalore** [comment]
@OlegHahm agree :) I like it much better than providing a size of 1

**Comment by OlegHahm** [comment]
Updated

**Comment by BytesGalore** [comment]
hmm, it complains the 0 size in `gnrc_networking` example and the `zep` test.
At least this shouldn't happen for the example since we use `MODULE_GNRC_IPV6` there, maybe related to #3256.
For the zep we need to specify a size though.

**Comment by BytesGalore** [comment]
OT (a bit): facing #3256 I'm getting a bit nervous regarding the FIB and passing `tables[]`:

For instance when we have 2 distinct modules defining `FIB_MAX_FIB_TABLE_ENTRIES`, say `mod_a` with 30 and `mod_b` with 10, we will probably iterate above the array bounds of `mod_b table[]`, and skip the last 10 `table[]` entries of `mod_a`.

**Comment by OlegHahm** [comment]
I don't get your last comment. If we have two modules using their own instance of FIB they should have two different constants for the size of "their" FIB.

**Comment by BytesGalore** [comment]
yes, but iff #3256 happens, the FIB uses a possibly wrong `FIB_MAX_FIB_TABLE_ENTRIES` for iteration, for instance the default one of `20`.

**Comment by OlegHahm** [comment]
Ah, yes, you're right, forgot to add a size parameter. Shame on me. Will update.

**Comment by OlegHahm** [comment]
See #3706.

**Comment by OlegHahm** [comment]
rebased to master

File: sys/include/universal_address.h [hunk 3]
```diff
@@ -28,7 +28,18 @@ extern "C" {
...
+                                            trigger warnings */
+#endif
+
+/* IPv6 address has 128 bit -> 16 bytes */
+#if defined(MODULE_GNRC_IPV6) && ((IPV6_ADDR_BIT_LEN >> 3) > UNIVERSAL_ADDRESS_SIZE)
```
Comment (cgundogan) [inline sys/include/universal_address.h]: to be more GNRC agnostic: you could use `MODULE_IPV6_ADDR` here instead.

**Comment by OlegHahm** [comment]
addressed comment

**Comment by OlegHahm** [comment]
should be fixed

File: sys/universal_address/universal_address.c [hunk 4]
```diff
@@ -28,7 +34,17 @@
 /**
  * @brief Maximum number of entries handled
  */
-#define UNIVERSAL_ADDRESS_MAX_ENTRIES (40)
+/* determine the maximum numer of entries */
+#ifndef UNIVERSAL_ADDRESS_MAX_ENTRIES
```
Comment (cgundogan) [inline sys/universal_address/universal_address.c]: @OlegHahm What's your opinion on moving this `ifdefing` into a separate file and `#include` it here? Maybe this kind of `ifdef`ing can be reused somewhen somewhere else..  It also would keep this `.c` file clean and you wouldn't need to include the headers at the very top

Comment (OlegHahm) [inline sys/universal_address/universal_address.c]: In general, sounds like a good idea, but for now I would leave it like this and change it once there is a need for this.

**Comment by cgundogan** [comment]
OK this seems to compile now. ACK

**Comment by OlegHahm** [comment]
squashed - but haven't tested the latest version

**Comment by cgundogan** [comment]
I quickly tested on native and configuring some entries in the fib. Nothing is broken so far

**Comment by OlegHahm** [comment]
Defined `UNIVERSAL_ADDRESS_MAX_ENTRIES` via unittest Makefile and squashed immediately.

**Pull request merged**